File: Python/Single/Single.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class SinglyLinkedList:
    def __init__(self):
        self.head = None

    def insert(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node

    def delete_node(self, key):
        temp = self.head
        prev = None

        if temp and temp.data == key:
            self.head = temp.next
            return

        while temp and temp.data != key:
            prev = temp
            temp = temp.next

        if temp is None:
            return

        prev.next = temp.next

    def delete_head(self):
        if self.head:
            self.head = self.head.next

    def delete_last(self):
        if self.head is None:
            return
        if self.head.next is None:
            self.head = None
            return
        cur = self.head
        while cur.next.next is not None:
            cur = cur.next
        cur.next = None

    def display(self):
        cur = self.head
        while cur:
            print(f"{cur.data} -> ", end="")
            cur = cur.next
        print("null")


if __name__ == "__main__":
    lst = SinglyLinkedList()
    lst.insert(3); lst.insert(2); lst.insert(1)
    print("Original List:")
    lst.display()
    lst.delete_node(2)
    print("After Deleting 2:")
    lst.display()

File: Java/Single/Single.java
public class SinglyLinkedList {
    static class Node {
        int data;
        Node next;
        Node(int d) { data = d; }
    }

    private Node head;

    public void insert(int data) {
        Node n = new Node(data);
        n.next = head;
        head = n;
    }

    public void deleteNode(int key) {
        Node temp = head, prev = null;

        if (temp != null && temp.data == key) {
            head = temp.next;
            return;
        }

        while (temp != null && temp.data != key) {
            prev = temp;
            temp = temp.next;
        }
        if (temp == null) return;

        prev.next = temp.next;
    }

    public void deleteHead() {
        if (head != null) head = head.next;
    }

    public void deleteLast() {
        if (head == null) return;
        if (head.next == null) { head = null; return; }
        Node cur = head;
        while (cur.next.next != null) cur = cur.next;
        cur.next = null;
    }

    public void display() {
        Node cur = head;
        while (cur != null) {
            System.out.print(cur.data + " -> ");
            cur = cur.next;
        }
        System.out.println("null");
    }

    public static void main(String[] args) {
        SinglyLinkedList list = new SinglyLinkedList();
        list.insert(3);
        list.insert(2);
        list.insert(1);

        System.out.println("Original List:");
        list.display();

        list.deleteNode(2);
        System.out.println("After Deleting 2:");
        list.display();
    }
}
